src/org/intellij/erlang/completion/QuoteInsertHandler.java

```diff
@@ -38,53 +35,55 @@ public QuoteInsertHandler(@NotNull String name) {
 
   @Override
   public void handleInsert(@NotNull InsertionContext context, LookupElement item) {
-    Project project = context.getProject();
-    process(project, myName, context);
+    process(myName, null, context);
     if (needColon()) {
-      AutoPopupController.getInstance(project).autoPopupMemberLookup(context.getEditor(), null);
+      AutoPopupController.getInstance(context.getProject()).autoPopupMemberLookup(context.getEditor(), null);
       super.handleInsert(context, item);
     }
   }
 
-  public static void process(@NotNull Project project, @NotNull String name, @NotNull InsertionContext context) {
-    process(project, name, null, context);
+  public static void process(@NotNull String name, @Nullable String moduleName, @NotNull InsertionContext context) {
+    if (moduleName != null && ErlangPsiImplUtil.needQuotation(moduleName)) {
+      int startOffset = context.getStartOffset();
+      insertQuotesIfAbsent(context, startOffset, startOffset + moduleName.length());
+    }
+    if (ErlangPsiImplUtil.needQuotation(name)) {
+      int tailOffset = context.getTailOffset();
+      int startOffset = moduleName == null ? context.getStartOffset() : tailOffset - name.length();
+      tailOffset += insertQuotesIfAbsent(context, startOffset, tailOffset);
+      context.getEditor().getCaretModel().moveToOffset(tailOffset);
+      context.setTailOffset(tailOffset);
+    }
   }
 
-  public static void process(@NotNull Project project, @NotNull String name, @Nullable String moduleName, @NotNull InsertionContext context) {
-    if (needQuotation(project, moduleName)) {
-      Editor editor = context.getEditor();
-      Document document = editor.getDocument();
+  private static int insertQuotesIfAbsent(@NotNull InsertionContext context, int startOffset, int tailOffset) {
+    int insertedQuotesCount = 0;
+    Document document = context.getEditor().getDocument();
+    CharSequence fileText = document.getCharsSequence();
+
+    if (!alreadyHasQuote(fileText, startOffset - 1)) {
       context.commitDocument();
-      int startOffset = context.getStartOffset();
       document.insertString(startOffset, QUOTA);
-      document.insertString(startOffset + moduleName.length() + 1, QUOTA);
+      insertedQuotesCount++;
+      tailOffset++;
     }
-    if (needQuotation(project, name)) {
-      Editor editor = context.getEditor();
-      Document document = editor.getDocument();
+    if (!alreadyHasQuote(fileText, tailOffset)) {
       context.commitDocument();
-      int tailOffset = context.getTailOffset();
-      int startOffset = moduleName == null ? context.getStartOffset() : tailOffset - name.length();
-      document.insertString(startOffset, QUOTA);
-      document.insertString(tailOffset + 1, QUOTA);
-      editor.getCaretModel().moveToOffset(tailOffset + 2);
-      context.setTailOffset(tailOffset + 2);
+      document.insertString(tailOffset, QUOTA);
+      insertedQuotesCount++;
```

Comment:
`-behaviour('<caret>')`
User wants to print here `gen_server`. He knows what he does and he have put these quotes not just because. Your idea of code will consume these quotes and it is not correct.